### ISSUE TITLE
Yummly prediction fails on large data set on Swan

### ISSUE THREAD

**Issue by buddha314** [buddha314/yummly-cuisine#1]
Details found on swan.

**Comment by ben-albrecht** [comment]
@bradcray ponders:

>   If there were a slow leak in [the] program, could it manifest in this way when run on larger problem sizes?

I ran locally with training sizes: 354 and 570 recipes. In both cases, the leaks were `O(1)` w.r.t. number of recipes.

See: [`--memLeaks` output](https://gist.github.com/ben-albrecht/7d3314b7a709925363edc0209f119ac7)

**Comment by buddha314** [comment]
Is that good or bad?

**Comment by ben-albrecht** [comment]
Good that there are no major leaks, and that we have one less thing to consider as the cause of the crash.

**Comment by gbtitus** [comment]
I reran this on swan yesterday evening/overnight.  I just started with a copy of the dir from the original failure, and made some changes to get additional information to try to see what was happening when it failed.  I added a '-' on the front of the `./$(EXEC) ...` line in `Makefile` so that it would keep running and do the following `date` command when the program terminated or was killed.  And while the job was running I used `pcmd()` to run `ps -u <myUser> v` to get some memory info every 10 seconds.

**edit:** I forgot to mention when I first wrote this up: I also redirected `stderr` in addition to `stdout` when I ran the `make(1)` command that ran the program, so I wouldn't miss any errors.  However, it's worth nothing that the output announcing the walltime timeout and job death did _not_ appear in that logfile, apparently because that output was sent to the `qsub` job rather than the `make(1)` which was running within it.  I'm considering running my `qsub` job within a `script(1)` session to try to capture absolutely everything, the next time I try this.

Unfortunately, the program failed only when it ran through the 12-hour walltime limit I'd set and the job was killed.  My understanding had been that the original run had failed after less than 12 walltime hours.  Was that the case?  If so, it's curious that mine didn't fail in that amount of time.

The last memory info I have before the program was killed shows TRS=7697k, DRS=3448914k, and RSS=55284k.  I'm not sure why TRS+DRS doesn't equal RSS, but ignoring that, this still isn't very much memory, nowhere near what's available on the nodes.  So we weren't about to OOM when the 12 hours ran out, at least.  For future runs I may try to get a better handle on memory usage, perhaps from /proc.

Also, the original log has content after the "loading graph" line, starting with "time to load matrix: 2814.04".  Is that time in seconds?  Interestingly, my run didn't produce any output after "loading graph" even though it ran for many hours after printing that.  However, it might have produced output that was still buffered inside the program and subsequently lost when the job was terminated with a signal due to the walltime limit.

Do we have some idea based on scaling runs how long the job should take on swan?  Also, do we know what kind of node the original swan run was on?  I didn't specify a node type when I ran and I also unfortunately didn't record which node (or type) I got, and furthermore I can't find that information in the Moab/Torque logs I can see.  But for future runs it might be good to use the same node type as the original run was done on, just for reproducibility's sake.

**Comment by ben-albrecht** [comment]
> Also, do we know what kind of node the original swan run was on? 

I assume the default batch which is the broadwell nodes.

Some local prediction timings:

| Total Ingredients in Recipes       | Time (s)      |
| ------------- |:-------------:|
| `285`           | `2.62857`      |
| `354`           | `5.39161`       |
| `469`           | `15.5569`       |
| `570`           | `40.7733`       |

Full size of `train.json` is `6714 recipes`


Here are 2 prediction timings I have from doing runs with `--memLeaks` (signficantly slower):

`354 : 286.171 seconds`
`570 : 1791.95 seconds`

**Comment by gbtitus** [comment]
Yeah, if you could do 2 or, better yet, 3 non-`memleaks` timings that would be cool.  It would be okay if the added one was smaller; I'm just curious about the O() of the main algorithm driving the time.  Or maybe someone could just tell me that off the top of their head, hint, hint?   :-)

**Comment by gbtitus** [comment]
I charted your non-`memLeaks` timings and found that `#recipes` vs. `ln(time)` looks pretty linear:

    ln(time) = (0.00962 * #recipes) - 1.78

Unfortunately for `#recipes=6714` I get `ln(time)=63` or so, and I don't think we can run for `e**63` seconds!

**Comment by ben-albrecht** [comment]
Perhaps the next step is to make things scale better then...

Are we currently suspecting the initial mysterious crash was a timeout that was not caught due to `stderr` not being piped?

**Comment by gbtitus** [comment]
I'm not sure, actually.  A walltime timeout produces pretty obvious output.  Here's what mine said last night:

    =>> PBS: job killed: walltime 43231 exceeded limit 43200
    Terminated

But there are ways to prevent this from being produced or if it is produced, seen, so I can't rule it out.

**Comment by buddha314** [comment]
I just chatted with Brad and I told him I'm "pretty sure" it died in about 5-7 hours, before the wall time limit.  However, I could be wrong.  I thought I would get a timed message if that happened.

Also, the "small" data set was built by taking a `head -n 2000` of the large data set, then closing the JSON format. Depending on where `head` lands, you would probably need to add `} ]` to the end of the file and remove extra commas.

**Comment by gbtitus** [comment]
I'm going to rerun the size=570 job (which @ben-albrecht tells me is the `head -n 2000` size) and capture memory information while that runs.  That will at least give us a better idea of the memory loading and whether OOM is likely/possible.  I hope I can get that done today still.

**Comment by gbtitus** [comment]
It suddenly occurs to me to ask: was a `craype-hugepages` module loaded during the original run?

**Comment by buddha314** [comment]
I would not know how to do that.



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------

**Comment by gbtitus** [comment]
Okay, it wasn't, then.  And it would only have had a minimal effect anyway, since this isn't a multi-locale program.

**Comment by buddha314** [comment]
Can it be made a multi locale program? Would that help?



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------

**Comment by gbtitus** [comment]
my `size=570` job finished in only a few minutes, with a memory high water mark at about 6 gb.  I'm inclined to think the program isn't memory constrained.

**Comment by gbtitus** [comment]
It could indeed.  A major feature of Chapel is the ease with which one can go from running on a single compute node to multiple compute nodes.  It wouldn't solve our problem, though, if indeed the algorithm is exponential-time.  We'll need to find a polynomial-time algorithm before it will be worthwhile to explore multi-locale execution.  I'll leave that to @ben-albrecht, who is the Chapel-and-apps expert for this issue.  I'm better at system stuff.

**Comment by buddha314** [comment]
I suck at that bit. Can we find the non linearities? I may be able to re write parts as vector ops



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------

**Comment by gbtitus** [comment]
Vectorization is usually beneficial, but it won't have much effect here.  The problem is that the program's running time is currently proportional to `e**W`, where `W` is the amount of work to be done.  At a high level, vectorization can reduce execution time by a factor of up to the vector width (typically at most 8 on these systems) and parallelization, including using multiple locales in order to increase the core count, can reduce execution time by a factor of up to the number of cores when the algorithm is highly parallel. But while those can be pretty nice improvements, especially the latter one, they're still easily overwhelmed by increases in the amount of work to be done.  For example, merely doubling the amount of work will effectively square the program's running time.  The way to solve that problem is to find a better algorithm, perhaps a so-called _polynomial-time_ one whose running time is proportional to `W**p` where `p` is > 1 but small (2 or 3), or better yet one with running time proportional to `W * log(W)`.  (A _linear_ algorithm with running time proportional to `W` would be best of all, but I strongly suspect there is no linear solution to the problem the program solves.)

**Comment by buddha314** [comment]
Point taken.  I'm not a computer scientist, but I play one on TV.  No, not even that.   The method I'm using is pretty complicated and has multiple parts, so I may be able to cut it down.  I don't know which algs are O(anything) so if someone can point out the expensive bits, I can look into those.
